plugins/linux-pipewire/camera-portal.c

```diff
@@ -0,0 +1,594 @@
+				     camera_source->device_id);
+
+	if (!device)
+		return;
+
+	blog(LOG_INFO, "[camera-portal] selected camera '%s'",
+	     camera_source->device_id);
+
+	camera_source->obs_pw_stream = obs_pipewire_connect_stream(
+		connection->obs_pw, camera_source->source, device->id,
+		"OBS PipeWire Camera",
+		pw_properties_new(PW_KEY_MEDIA_TYPE, "Video",
+				  PW_KEY_MEDIA_CATEGORY, "Capture",
+				  PW_KEY_MEDIA_ROLE, "Camera", NULL));
+}
+
+static bool device_selected(void *data, obs_properties_t *props,
+			    obs_property_t *property, obs_data_t *settings)
+{
+	struct camera_portal_source *camera_source = data;
+	const char *device_id;
+
+	device_id = obs_data_get_string(settings, "device_id");
+
+	if (update_device_id(camera_source, device_id))
+		stream_camera(camera_source);
+
+	return true;
+}
+
+static void populate_cameras_list(struct camera_portal_source *camera_source,
+				  obs_property_t *device_list)
+{
+	struct camera_device *device;
+	GHashTableIter iter;
+	const char *device_id;
+	bool device_found;
+
+	if (!connection)
+		return;
+
+	obs_property_list_clear(device_list);
+
+	device_found = false;
+	g_hash_table_iter_init(&iter, connection->devices);
+	while (g_hash_table_iter_next(&iter, (gpointer *)&device_id,
+				      (gpointer *)&device)) {
+		const char *device_name;
+
+		device_name = pw_properties_get(device->properties,
+						PW_KEY_NODE_DESCRIPTION);
+
+		obs_property_list_add_string(device_list, device_name,
+					     device_id);
+
+		device_found |= strcmp(device_id, camera_source->device_id) ==
+				0;
+	}
+
+	if (!device_found && camera_source->device_id) {
+		size_t device_index;
+		device_index = obs_property_list_add_string(
+			device_list, camera_source->device_id,
+			camera_source->device_id);
+		obs_property_list_item_disable(device_list, device_index, true);
+	}
+}
+
+/* ------------------------------------------------- */
+
+static void on_proxy_removed_cb(void *data)
+{
+	struct camera_device *device = data;
+	pw_proxy_destroy(device->proxy);
+}
+
+static void on_destroy_proxy_cb(void *data)
+{
+	struct camera_device *device = data;
+
+	spa_hook_remove(&device->proxy_listener);
+
+	device->proxy = NULL;
+}
+
+static const struct pw_proxy_events proxy_events = {
+	PW_VERSION_PROXY_EVENTS,
+	.removed = on_proxy_removed_cb,
+	.destroy = on_destroy_proxy_cb,
+};
+
+static void on_registry_global_cb(void *user_data, uint32_t id,
+				  uint32_t permissions, const char *type,
+				  uint32_t version,
+				  const struct spa_dict *props)
+{
+	UNUSED_PARAMETER(user_data);
+	UNUSED_PARAMETER(permissions);
+	UNUSED_PARAMETER(version);
+
+	struct camera_device *device;
+	struct pw_registry *registry;
+	const char *device_id;
+
+	if (strcmp(type, PW_TYPE_INTERFACE_Node) != 0)
+		return;
+
+	registry = obs_pipewire_get_registry(connection->obs_pw);
+	device_id = spa_dict_lookup(props, SPA_KEY_NODE_NAME);
+
+	device = camera_device_new(id, props);
+	device->proxy = pw_registry_bind(registry, id, type, version, 0);
```

Comment:
Is there a reason, why we create a proxy for each PipeWire? I think we don't need that.

Reply:
This is what pw-dump does, and apparently tracks the life of each device individually